### PR TITLE
[GEOS-9343] Update data rules (layer and layer groups) rules at runtime

### PR DESCRIPTION
**Pull request by taba90**
<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [X] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [X] PR for bug fixes and small new features are presented as a single commit
- [X] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [X] New unit tests have been added covering the changes
- [X] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [X] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by nmco** [comment]
JIRA ticket:
https://osgeo-org.atlassian.net/browse/GEOS-9343

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java [hunk 1]
```diff
@@ -1526,6 +1531,36 @@ public void testRemoveLayer() {
         assertTrue(catalog.getLayers().isEmpty());
     }
 
+    @Test
+    public void testRemoveLayerAndAssociatedDataRules() throws IOException {
+        DataAccessRuleDAO dao = DataAccessRuleDAO.get();
+        CatalogListener listener = new SecuredResourceNameChangeListener(catalog, dao);
+        // catalog.getListeners().add()
```
Comment (nmco) [inline src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java]: Let over comment?

File: src/main/src/main/java/org/geoserver/security/SecuredResourceNameChangeListener.java [hunk 2]
```diff
@@ -50,7 +45,31 @@ public void handleAddEvent(CatalogAddEvent event) throws CatalogException {
...
+            resourceName = ((LayerInfo) resourceInfo).getName();
+        } else if (resourceInfo != null && event.getSource() instanceof LayerGroupInfo) {
+            resourceWorkSpaceName = ((LayerGroupInfo) resourceInfo).getWorkspace().getName();
+            resourceName = ((LayerGroupInfo) resourceInfo).getName();
+        }
+
+        if (resourceWorkSpaceName != null && resourceInfo != null) {
+            // if a security rule exists for the layer whose name has been changed
+            if (layerHasSecurityRule(resourceWorkSpaceName, resourceName)) {
+                LOGGER.info(
```
Comment (nmco) [inline src/main/src/main/java/org/geoserver/security/SecuredResourceNameChangeListener.java]: Nice touch!

File: src/main/src/main/java/org/geoserver/security/SecuredResourceNameChangeListener.java [hunk 3]
```diff
@@ -50,7 +45,31 @@ public void handleAddEvent(CatalogAddEvent event) throws CatalogException {
 
     @Override
     public void handleRemoveEvent(CatalogRemoveEvent event) throws CatalogException {
-        // ignore
+        String resourceWorkSpaceName = null;
+        CatalogInfo resourceInfo = event.getSource();
+        String resourceName = null;
+        if (resourceInfo != null && resourceInfo instanceof LayerInfo) {
```
Comment (nmco) [inline src/main/src/main/java/org/geoserver/security/SecuredResourceNameChangeListener.java]: If I'm not mistaken, ``resourceInfo instanceof LayerInfo`` will yield ``false`` if ``resourceInfo`` is NULL, hence no need the ``resourceInfo != null``check.

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java [hunk 4]
```diff
@@ -2149,6 +2184,31 @@ public void testGetLayerGroupByName() {
         assertNull(catalog.getLayerGroupByName("cite", "layerGroup2"));
     }
 
+    @Test
+    public void testRemoveLayerGroupAndAssociatedDataRules() throws IOException {
+        DataAccessRuleDAO dao = DataAccessRuleDAO.get();
+        CatalogListener listener = new SecuredResourceNameChangeListener(catalog, dao);
+        // catalog.getListeners().add()
```
Comment (nmco) [inline src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java]: ?

**Comment by nmco** [comment]
Just some minor comments, manually tested this seems to be working fine :+1:

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java (hunk 4 above)
Comment (taba90) [reply src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java]: do you mean that is not necessary to test layer group also?

**Comment by nmco** [comment]
Great thx @taba90 !

**Pull request merged**